src/python/pants/backend/terraform/dependency_inference.py

```diff
@@ -0,0 +1,186 @@
+from pants.engine.fs import CreateDigest, Digest, FileContent
+from pants.engine.internals.selectors import Get
+from pants.engine.process import Process, ProcessResult
+from pants.engine.rules import collect_rules, rule
+from pants.engine.target import (
+    HydratedSources,
+    HydrateSourcesRequest,
+    InferDependenciesRequest,
+    InferredDependencies,
+    Targets,
+)
+from pants.engine.unions import UnionRule
+from pants.util.ordered_set import OrderedSet
+
+PARSER = FileContent(
+    "__pants_tf_parser.py",
+    textwrap.dedent(
+        """\
+    from pathlib import PurePath
+    import sys
+    from typing import Set
+
+    import hcl2
+
+    # PurePath does not have the Path.resolve method which resolves ".." components, thus we need to
+    # code our own version for PurePath's.
+    def resolve_pure_path(base: PurePath, relative_path: PurePath) -> PurePath:
+        parts = list(base.parts)
+        for component in relative_path.parts:
+            if component == ".":
+                pass
+            elif component == "..":
+                if not parts:
+                    raise ValueError(f"Relative path {relative_path} escapes from path {base}.")
+                parts.pop()
+            else:
+                parts.append(component)
+
+        return PurePath(*parts)
+
+
+    def extract_module_source_paths(path: PurePath, raw_content: bytes) -> Set[str]:
+        content = raw_content.decode("utf-8")
+        parsed_content = hcl2.loads(content)
+
+        # Note: The `module` key is a list where each entry is a dict with a single entry where the key is the
+        # module name and the values are a dict for that module's actual values.
+        paths = set()
+        for wrapped_module in parsed_content.get("module", []):
+            values = list(wrapped_module.values())[
+                0
+            ]  # the module is the sole entry in `wrapped_module`
+            source = values.get("source", "")
+
+            # Local paths to modules must begin with "." or ".." as per
+            # https://www.terraform.io/docs/language/modules/sources.html#local-paths.
+            if source.startswith("./") or source.startswith("../"):
+                try:
+                    resolved_path = resolve_pure_path(path, PurePath(source))
+                    paths.add(str(resolved_path))
+                except ValueError:
+                    pass
+
+        return paths
+
+
+    paths = set()
+    for filename in sys.argv[1:]:
+        with open(filename, "rb") as f:
+            content = f.read()
+        paths |= extract_module_source_paths(PurePath(filename).parent, content)
+
+    for path in paths:
+        print(path)
+    """
+    ).encode("utf-8"),
+    is_executable=True,
+)
+
+
+class PythonHcl2Subsystem(PythonToolRequirementsBase):
+    options_scope = "python-hcl2"
+    help = "Used to parse Terraform modules to infer their dependencies."
+
+    default_version = "python-hcl2==3.0.1"
+
```

Comment:
This should define a lockfile by setting https://github.com/pantsbuild/pants/blob/25d64887cb93f5abd7dea6f6029abd9b5927ee20/src/python/pants/backend/python/subsystems/python_tool_base.py#L39-L48 ... I don't know what the right values are, but @Eric-Arellano / @chrisjrn can chime in.

Reply:
The example that @Eric-Arellano pointed me at for poetry does not do that though.

@Eric-Arellano  / @chrisjrn: What is the right thing to do here?

Reply:
The only reason Poetry doesn't have it is it would be a chicken-and-egg issue with bootstrapping. We use Poetry to generate Poetry lockfiles. Solvable, but we didn't do it because Poetry is going away in favor of Pex/pip.

--

Indeed, please do support lockfiles here. See Docformatter as a simple example:

https://github.com/pantsbuild/pants/blob/682828e954bb2507afb8dc7ccc19d97376a14cfa/src/python/pants/backend/python/lint/docformatter/subsystem.py#L25-L28

The lockfile should probably live in `backend/terraform/hcl2_lockfile.txt`.

A rule to register this with `./pants generate-lockfiles`:

https://github.com/pantsbuild/pants/blob/682828e954bb2507afb8dc7ccc19d97376a14cfa/src/python/pants/backend/python/lint/docformatter/subsystem.py#L61-L73

Tests that we support this lockfile on macOS too, and all expected Python versions:

https://github.com/pantsbuild/pants/blob/682828e954bb2507afb8dc7ccc19d97376a14cfa/src/python/pants/backend/python/lint/docformatter/rules_integration_test.py#L75-L92

You only need that for one test to prove it works, not your `test_resolve_pure_path` for example. The `python_tests` target needs a tag:

https://github.com/pantsbuild/pants/blob/682828e954bb2507afb8dc7ccc19d97376a14cfa/src/python/pants/backend/python/lint/docformatter/BUILD#L7-L13

Finally, update our script to generate lockfiles:

https://github.com/pantsbuild/pants/blob/682828e954bb2507afb8dc7ccc19d97376a14cfa/build-support/bin/_generate_all_lockfiles_helper.py#L116-L121

Run `build-support/bin/generate_all_lockfiles.sh` to generate it. (You can optionally revert any unrelated lockfiles that get changed - we need to improve this script still so you can tell it to just generate a single lockfile.)